README.md
# deborg

**deborg** is a small utility script to extract .`deb` packages from an emacs
`org`mode file.

When setting up a new machine, one big task is to get all the functionality back
that one relied on for productivity on the old system, which also entails
installing the same packages. Furthermore, when using several computers with
slightly different setup requirements (one Debian, one Ubuntu, another a netbook
which can't handle everything), this can become time consuming.

To be able to get a newly installed system into the state I want it in, I
started documenting all packages that I install on top of the default: for which
purpose I need them; if there where any quirks, and whatever notes might be
useful. For this I use a simple [orgmode][orgmode] file in which every section
refers to a task / functionality and a list in which all the packages needed for
this functionality are mentioned.

**deborg** can then be used to easily extract all packages for a system based on
distro and release as an array, which can then be passed to e.g. apt to do all
the work.

```
orgfile="my_package_file.org"
distro=$(lsb_release --short --id)
release=$(lsb_release --short --release)

packages_to_install=($(deborg "$orgfile" "$distro" "$release"))
apt install "${packages_to_install[@]}"
```

Done! New system has all specified packages installed (if they are available).


## Installation


## Expected org file input

The orgmode file can be written as a normal file with sections, subsections,
markup (~code~, =verbatim=, ...), and text (only tables may be a problem).

However, each list entry (only entries using `+` or `-` as bullets) will be
considered to contain information about packages. Each package line can contain
any number of comma separated *alternatives* - this means, from each line (=
list entry) at maximum one package will be returned. The choice here is from
most-specific-fit to least-specific-fit with regards to distro and release: If
an exact fit (distro & release) is found, this will be returned, otherwise if a
package is defined that has no release specified and matches the specified
distro, this will be returned, lastly if the line contains a package without any
further specification, this will be returned (a package will never be returned
if distro or release mis-match). Examples:

---

* `+ package` - 
  install package for all distros and all releases
* `+ package {Debian}` -
  install package *only* if distro=Debian
* `+ package1, package2 {Ubuntu}` -
  install package1 for *all* distros, **unless** distro=Ubuntu, then *don't*
  install package1 but package2 *instead*
* `+ package1 {Debian}, package2 {Ubuntu:18.04}` -
  install package1 for Debian, and *only for Ubuntu 18.04* install *package2*; for
  other releases of Ubuntu *no* package will be installed (returned by parsing
  the orgfile)

---

Example of a valid orgfile:


```
# File listing all packages that should be installed on a linux system
#+LANGUAGE: en
#+TITLE: Linux Packages

* Emacs
  These are some packages needed for the emacs config to work, and yes, comments
  like this one are perfectly ok, as are links as seen below.

** elpy ([[https://github.com/jorgenschaefer/elpy][github]])
   + python3-jedi
   + black
   + python3-autopep8
   + yapf3
   + python3-yapf
   + virtualenv

* Email
  + thunderbird
    
** Languages
  + thunderbird-l10n-en-gb {Debian}, thunderbird-locale-en-gb {Ubuntu} :: british english
  + thunderbird-l10n-de {Debian}, thunderbird-locale-de {Ubuntu} :: german
  + packagex-here :: this type of comment is ok
  
** Other
  Not sure about these packages, but want to keep them here for now, even if
  not installing them by commenting them out.
  # + package_not_returned :: as the line does not start with either - or +
  
```


## Output

**deborg** returns all extracted packages as a string separated by the specified
**sep**arator which defaults to space `' '`.


## Examples / Usage

Input orgmode file:
```
# this is the input file for the example (example.org)
* Some packages
  - package1
  - package2a {Debian}, package2b {Debian:11}, package2c {Ubuntu}
  - package3 {Debian}
  - package4 {Debian}, package4 {Ubuntu:20.04}
```

Output:

Normally use `lsb_release` in a script to obtain *'distro'* and *'release'*, we
set it here explicitly for clarity.

1) on a Debian 10 system
```
$ deborg example.org "Debian" "10"
package1 package2a package3 package4
$
```

2) on a Debian 11 system
```
$ deborg example.org "Debian" "11"
package1 package2b package3 package4
$
```

3) on a Ubuntu 18.04 system
```
$ deborg example.org "Ubuntu" "18.04"
package1 package2c
$
```

4) on a Ubuntu 20.04 system
```
$ deborg example.org "Ubuntu" "20.04"
package1 package2c package4
$
```

5) setting **sep**
```
$ deborg example.org "Debian" "9" --sep=":::"
package1:::package2a:::package3:::package4
$
```


## License

**deborg** is released under the [GPL-3][GPL-3 spdx].


[orgmode]: https://orgmode.org/
[GPL-3 spdx]: https://spdx.org/licenses/GPL-3.0-or-later.html


debian/changelog
deborg (1.0.0-1) unstable; urgency=low

  * initial release

 -- Tobias Marczewski (mtoboid) <vortex@e.mail.de>  Fri, 04 Mar 2022 16:42:00 +0100




tests/test_deborg.py
"""
Tests for the command line utility deborg accessed via deborg.__main__.py.
"""

from __future__ import annotations

from pytest_console_scripts import RunResult

from deborg.cli import Examples


def test_two_package_match_in_same_line_raises_error(tmpdir, script_runner):
    file_content: str = "" +\
        "# file created by pytest test_deborg.py\n" +\
        "# this item entry contains two packages matching any distro and release:\n" +\
        "+ package-1-a, package-1-b\n"
    error_msg_start: str = "Error while parsing"
    error_msg_end: str = "Error in line 2: More than two packages match " +\
                         "the specifications: package-1-a, package-1-b."
    orgfile = tmpdir.join("testfile.org")
    orgfile.write(file_content)
    result = script_runner.run('deborg', str(orgfile), 'distro', 'release')
    assert result.returncode == 1
    assert result.stderr.strip().startswith(error_msg_start)
    assert result.stderr.strip().endswith(error_msg_end)


def test_version_argument_works(script_runner):
    result = script_runner.run('deborg', '--version')
    assert result.success
    assert result.stderr == ''


def test_non_existing_orgfile_raises_error(script_runner):
    """Try to open a non-existing file."""
    result = script_runner.run(
        'deborg', '/not/existing/file.org',
        'any_distro', 'any_release')
    assert not result.success
    assert result.returncode == 1
    assert result.stdout.strip().endswith("not found.")


def test_parse_testfile(script_runner):
    """Parse the test file with one setting for distro and repo."""
    expected_packages: list[str] = [
        "basic-commented-package2",
        "basic-package1",
        "package",
        "package2",
        "package5-hyphen",
        "package-distA11",
        "package-distA12",
        "package-distA5",
        "package-distA6",
        "package-with-hyphens",
        "packageX"]
    result = script_runner.run(
        'deborg', 'tests/input/testfile_ex1.org',
        'distroA', 'release0', '--sep=::')
    assert result.success
    output: list[str] = result.stdout.strip().split("::")
    # sort both lists as shell 'sort' seems to sort '-' differently
    assert sorted(output) == sorted(expected_packages)


def test_example_file_examples(tmpdir, script_runner):
    tests: Examples = Examples()
    orgfile = tmpdir.join("testfile.org")
    command_out = script_runner.run('deborg', '--example-file')
    orgfile.write(command_out.stdout)
    for i in range(tests.n):
        # the parsed input 'distro1' becomes "'distro1'" which is not "distro1"
        # -> remove all ' from the strings
        args = [a.replace("'", "") for a in tests.input[i].split(' ')]
        command = ['deborg', str(orgfile), *args]
        result: RunResult = script_runner.run(*command)
        assert result.returncode == 0
        assert result.stdout.strip() == tests.output[i].strip()


tests/test_parser.py
from __future__ import annotations

import pytest
from collections.abc import Sequence
from pathlib import Path

from deborg.orgparser import DebPakInfo, OrgParser


class TestDebPakInfo:
    """Tests for the dataclass used to hold deb package information."""
    def test_instantiation_name_only(self):
        d: DebPakInfo = DebPakInfo("package_name")
        assert d.name == "package_name"

    def test_instantiation_name_and_distro(self):
        d: DebPakInfo = DebPakInfo("package_name", distro="distro_name")
        assert (d.name == "package_name" and d.distro == "distro_name")

    def test_instantiation_name_distro_release(self):
        d: DebPakInfo = DebPakInfo(
            "package_name", distro="distro_name", release="release_name"
        )
        assert (
                d.name == "package_name"
                and d.distro == "distro_name"
                and d.release == "release_name"
        )

    def test_instantiation_with_tags(self):
        d: DebPakInfo = DebPakInfo(
            "package_name", distro="distro_name", release="release_name",
            tags=["tag1", "tag_2", "tag-3"]
        )
        assert (
                d.name == "package_name"
                and d.distro == "distro_name"
                and d.release == "release_name"
                and len(d.tags) == 3
                and d.tags == ["tag1", "tag_2", "tag-3"]
        )

    def test_equal_objects_are_equal(self):
        d1: DebPakInfo = DebPakInfo(
            "package1", distro="distroA", release="releaseA1",
            tags=["tag1", "tag2", "tag3"]
        )
        d2: DebPakInfo = DebPakInfo(
            "package1", distro="distroA", release="releaseA1",
            tags=["tag1", "tag2", "tag3"]
        )
        assert d1 == d2

    def test_unequal_objects_are_not_equal(self):
        d1: DebPakInfo = DebPakInfo(
            "package1", distro="distroA", release="releaseA1",
            tags=["tag1", "tag2", "tag3"]
        )
        d2: DebPakInfo = DebPakInfo(
            "package2", distro="distroB", release="releaseB1",
            tags=["tag1", "tag2", "tag3"]
        )
        assert d1 != d2


@pytest.fixture
def non_package_lines() -> Sequence[str]:
    """Text that could occur in the file, but is not a line specifying a deb package."""
    return [
        "",
        "This is just a line",
        "A line with some + and - symbols in it.",
        "One with a colon: will that pass?",
        "* A Heading - yep",
        "** Subheading",
        "The next word is =verbatim=",
        "Here we have a ~code~ tag.",
        "No idea when one would do {that}, but whatever."
    ]


@pytest.fixture
def package_info_str_input() -> Sequence[tuple[str, DebPakInfo]]:
    return [
        ("package",
         DebPakInfo(name="package")),
        (" pack1 {Debian} ",
         DebPakInfo(name="pack1", distro="Debian")),
        (" p-1-package {Debian:10}  ",
         DebPakInfo(name="p-1-package", distro="Debian", release="10")),
        (" thisisa-n20-package {Ubuntu:bionic}",
         DebPakInfo(name="thisisa-n20-package", distro="Ubuntu", release="bionic")),
        ("otherone {Ubuntu:18.04}",
         DebPakInfo(name="otherone", distro="Ubuntu", release="18.04")),
        ("package-tags-only {::one,two,three}",
         DebPakInfo(name="package-tags-only", tags=["one", "two", "three"])),
        ("tags-and-distro {Debian::one,and-two,and_three}",
         DebPakInfo(name="tags-and-distro", distro="Debian",
                    tags=["one", "and-two", "and_three"])),
        ("package-all {Ubuntu:20.04:netbook,laptop}",
         DebPakInfo(name="package-all", distro="Ubuntu", release="20.04",
                    tags=["netbook", "laptop"]))
    ]


@pytest.fixture
def package_line_split_input() -> Sequence[tuple[str, list[str]]]:
    return [
        ("+ package",
         ["package"]),

        ("- package1, package2 :: comment is ignored",
         ["package1", "package2"]),

        ("- package-1, package-1b {distro1::tag1} :: comment is ignored.",
         ["package-1", "package-1b {distro1::tag1}"]),

        (" + package-nn12 {dist:11:one,two}, pack2 {dist2}, p3 {dist2::three,f0ur} :: blurb",
         ["package-nn12 {dist:11:one,two}", "pack2 {dist2}", "p3 {dist2::three,f0ur}"]),

        (" + package {::one,xxx,m} :: one was for l, and others, well...",
         ["package {::one,xxx,m}"])
    ]


@pytest.fixture
def deb_package_line_input() -> Sequence[tuple[str, str, str, DebPakInfo]]:
    """1) Input line | 2) distro (arg) | 3) release (arg) | 4) expected out"""
    return [
        ("This is not a package line...",
         "any_distro", "any_release",
         None),

        ("This neither but it has a + and a - in it.",
         "any_distro", "any_release",
         None),

        (" + package1",
         "any_distro", "any_release",
         DebPakInfo(name="package1")),

        (" - package-x {Distro} ",
         "not_distro", "any_release",
         None),

        (" - package-x {Distro} ",
         "Distro", "any_release",
         DebPakInfo(name="package-x", distro="Distro")),

        ("+ pak-a {Debian}, pak-b {theOtherDistro:42}, pak-c ",
         "any_distro", "any_release",
         DebPakInfo(name="pak-c")),

        ("+ pak-a {Debian}, pak-b {theOtherDistro:42}, pak-c ",
         "Debian", "any_release",
         DebPakInfo(name="pak-a", distro="Debian")),

        ("+ pak-a {Debian}, pak-b {theOtherDistro:42}, pak-c ",
         "theOtherDistro", "any_release",
         DebPakInfo(name="pak-c")),

        ("+ pak-a {Debian}, pak-b {theOtherDistro:42}, pak-c ",
         "theOtherDistro", "42",
         DebPakInfo(name="pak-b", distro="theOtherDistro", release="42")),

        ("  - pakx :: this is a great package!",
         "any_distro", "any_release",
         DebPakInfo(name="pakx")),

        ("+ pak-all, pak-special {forSpecial:321} :: special needs special + love",
         "any_distro", "any_release",
         DebPakInfo(name="pak-all")),

        ("+ pak-all, pak-special {forSpecial:321} :: special needs special + love",
         "forSpecial", "other_release",
         DebPakInfo(name="pak-all")),

        ("+ pak-all, pak-special {forSpecial:321} :: special needs special + love",
         "forSpecial", "321",
         DebPakInfo(name="pak-special", distro="forSpecial", release="321")),
    ]


Tags = list[str]


@pytest.fixture
def deb_package_line_input_with_tags() ->\
        Sequence[tuple[str, str, str, Tags, DebPakInfo]]:
    """
    1 - Input line |
    2 - distro (arg) | 3 - release (arg) | 4 - tags (arg) |
    5 - expected output
    """
    return [
        ("+ package-no-tag-match {distro1:release1:tag1,tag2,tag3}",
         "distro1", "release1", ["tag4", "tag5"],
         None),

        ("+ package-single-tag-match {distro1:release1:tag2,tag3}",
         "distro1", "release1", ["tag2"],
         DebPakInfo(name="package-single-tag-match",
                    distro="distro1",
                    release="release1",
                    tags=["tag2", "tag3"])),

        ("+ package-one-tag-matches {distro1:release1:tag1,tag2}",
         "distro1", "release1", ["tag2", "tag5"],
         DebPakInfo(name="package-one-tag-matches",
                    distro="distro1",
                    release="release1",
                    tags=["tag1", "tag2"])),

        ("+ package-multiple-tag-match {distro1:release1:tag2,tag3,tag6,tag7}",
         "distro1", "release1", ["tag2", "tag7"],
         DebPakInfo(name="package-multiple-tag-match",
                    distro="distro1",
                    release="release1",
                    tags=["tag2", "tag3", "tag6", "tag7"])),

        ("+ package-no-tag",
         "distro", "release", ["tag3"],
         DebPakInfo(name="package-no-tag")),

        ("+ package-only-tag-match {::tag3}",
         "distro", "release", ["tag3"],
         DebPakInfo(name="package-only-tag-match",
                    tags=["tag3"])),

        ("+ package-only-tag-no-match {::tag3}",
         "distro", "release", [],
         None),

        ("+ package-tag-only-and-comment {::tag3} :: this should not cause problems.",
         "distro", "release", ["tag3"],
         DebPakInfo(name="package-tag-only-and-comment",
                    tags=["tag3"])),

        ("+ package-only-distro {distro1}",
         "distro1", "release1", ["tag1", "tag2"],
         DebPakInfo(name="package-only-distro",
                    distro="distro1")),

        ("+ package-distro-and-release {distro1:release1}",
         "distro1", "release1", ["tag1", "tag2"],
         DebPakInfo(name="package-distro-and-release",
                    distro="distro1",
                    release="release1")),

        ("+ package1 {distro1::tag1}, package2 {distro1::tag2}",
         "distro1", "release1", ["tag2"],
         DebPakInfo(name="package2",
                    distro="distro1",
                    tags=["tag2"])),

        ("+ package1 {distro1::tag1}, package2 {distro1::tag2}",
         "distro1", "release1", ["tag1"],
         DebPakInfo(name="package1",
                    distro="distro1",
                    tags=["tag1"])),

        ("- pk1 {A::tag1,tag2}, pk2 {B}, pk3 {B::tag2} :: some comment here.",
         "A", "release", [],
         None),

        ("- pk1 {A::tag1,tag2}, pk2 {B}, pk3 {B::tag2} :: some comment here.",
         "A", "release", ["tag2"],
         DebPakInfo(name="pk1",
                    distro="A",
                    tags=["tag1", "tag2"])),

        ("- pk1 {A::tag1,tag2}, pk2 {B}, pk3 {B::tag2} :: some comment here.",
         "B", "release", [],
         DebPakInfo(name="pk2",
                    distro="B")),

        ("- pk1 {A::tag1,tag2}, pk2 {B}, pk3 {B::tag2} :: some comment here.",
         "B", "release", ["tag2"],
         DebPakInfo(name="pk3",
                    distro="B",
                    tags=["tag2"])),

        ("- pk1 {A::tag1,tag2}, pk2 {B}, pk3 {B::tag2} :: some comment here.",
         "B", "release", ["tag1"],
         DebPakInfo(name="pk2",
                    distro="B"))
    ]


# packages obtained from parsing a file
ExpectedPackagesOutput = Sequence[str]
# distro, release, expected_output
InputOutputSpec = tuple[str, str, ExpectedPackagesOutput]


@pytest.fixture
def org_file1_tests() -> tuple[Path, Sequence[InputOutputSpec]]:
    """Parsing tests for file ./tests/input/testfile_ex1.org."""
    testfile: str = "input/testfile_ex1.org"
    test_dir: Path = Path(__file__).resolve().parent
    test_input_org_file: Path = test_dir.joinpath(testfile)
    if not test_input_org_file.exists():
        raise FileNotFoundError(f"Testfile: {testfile} not found.")

    test_params: list[InputOutputSpec] = [
        ("distro", "any",
         [
            "package",
            "package2",
            "package-with-hyphens",
            "package5-hyphen",
            "packageX",
            "basic-package1",
            "basic-commented-package2",
            "package-dist3",
            "package-dist4",
            "package-distA11"
         ]),

        ("distroA", "any",
         [
            "package",
            "package2",
            "package-with-hyphens",
            "package5-hyphen",
            "packageX",
            "basic-package1",
            "basic-commented-package2",
            "package-distA5",
            "package-distA6",
            "package-distA11",
            "package-distA12"
         ]),

        ("distroA", "release1",
         [
            "package",
            "package2",
            "package-with-hyphens",
            "package5-hyphen",
            "packageX",
            "basic-package1",
            "basic-commented-package2",
            "package-distA5",
            "package-distA6",
            "package-distA7",
            "package-distA8",
            "package-distA9",
            "package-distA10",
            "package-distA11",
            "package-distA12"
         ]),

        ("distroA", "release2",
         [
            "package",
            "package2",
            "package-with-hyphens",
            "package5-hyphen",
            "packageX",
            "basic-package1",
            "basic-commented-package2",
            "package-distA5",
            "package-distA6",
            "package-distB9",
            "package-distA11",
            "package-distB12"
         ]),

        ("distroB", "any",
         [
            "package",
            "package2",
            "package-with-hyphens",
            "package5-hyphen",
            "packageX",
            "basic-package1",
            "basic-commented-package2",
            "package-distB5",
            "package-distB6",
            "package-distA11"
         ]),

        ("distroB", "release1",
         [
            "package",
            "package2",
            "package-with-hyphens",
            "package5-hyphen",
            "packageX",
            "basic-package1",
            "basic-commented-package2",
            "package-distB5",
            "package-distB6",
            "package-distB11"
         ]),

        ("distroB", "release2",
         [
            "package",
            "package2",
            "package-with-hyphens",
            "package5-hyphen",
            "packageX",
            "basic-package1",
            "basic-commented-package2",
            "package-distB5",
            "package-distB6",
            "package-distB10",
            "package-distA11"
         ])
    ]
    return test_input_org_file, test_params


class TestParser:
    """Tests for the parser itself."""

    def test_get_package_info(self, package_info_str_input):
        for line, pak_info in package_info_str_input:
            assert OrgParser._get_package_info(line) == pak_info

    def test_split_package_line(self, package_line_split_input):
        for line, expected in package_line_split_input:
            assert OrgParser._split_package_line(line) == expected

    def test_line_containing_no_package_info_returns_empty_list(self, non_package_lines):
        for line in non_package_lines:
            output: DebPakInfo | None = OrgParser.extract_deb_package_from_line(line, "any", "any")
            assert output is None

    def test_extract_deb_package_from_line_no_tags(self, deb_package_line_input):
        for line, *param, expected_DebPakInfo in deb_package_line_input:
            package_info = OrgParser.extract_deb_package_from_line(line, *param)
            assert package_info == expected_DebPakInfo

    def test_extract_deb_package_from_line_w_tags(self, deb_package_line_input_with_tags):
        for line, *param, expected_DebPakInfo in deb_package_line_input_with_tags:
            package_info = OrgParser.extract_deb_package_from_line(line, *param)
            assert package_info == expected_DebPakInfo

    def test_extract_deb_packages(self, org_file1_tests):
        file: Path
        io_args: InputOutputSpec
        file, io_args = org_file1_tests
        for args in io_args:
            distro, release, expected_out = args
            packages: list[str] = OrgParser.extract_deb_packages(
                file, distro, release)
            assert sorted(packages) == sorted(expected_out)

src/deborg/__init__.py
# SPDX-License-Identifier: GPL-3.0-or-later
#
# Copyright (C) 2022 Tobias Marczewski
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""
deborg - A package to provide the shell command 'deborg' for parsing emacs orgmode
files for debian package information.

modules:
    __main__  : Provides the shell command via argparse.
    parser    : Contains the logic to parse orgfiles for debian package info.
"""


src/deborg/__main__.py
#!/usr/bin/python3
#
# SPDX-License-Identifier: GPL-3.0-or-later
#
# Copyright (C) 2022 Tobias Marczewski
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""Provides a command line entry point for the program deborg."""

from __future__ import annotations

__author__ = "Tobias Marczewski (mtoboid)"
__version__ = "1.0.0"

import sys

from argparse import ArgumentParser
from pathlib import Path

from deborg.cli import cli_parser
from deborg.orgparser import OrgParser, OrgParserError


def main():
    parser: ArgumentParser = cli_parser()
    args = parser.parse_args()
    file: Path = Path(args.orgfile)

    if not file.exists():
        print(f"Error: specified file '{file.resolve().as_posix()}' not found.")
        sys.exit(1)

    _tags: list[str] | None = None
    if args.tags:
        _tags: list[str] = args.tags.split(",")

    try:
        packages: list[str] = OrgParser.extract_deb_packages(file, args.distro, args.release, _tags)
        sys.stdout.write(args.sep.join(packages))
        sys.exit(0)
    except OrgParserError as pe:
        sys.stderr.write(f"Error while parsing {file.as_posix()}:\n{pe}")
        sys.exit(1)


if __name__ == '__main__':
    main()


src/deborg/cli.py
# SPDX-License-Identifier: GPL-3.0-or-later
#
# Copyright (C) 2022 Tobias Marczewski
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

"""
Provides the cli for the package 'deborg'.

Classes:

    PrintExampleFile
    Examples

Misc variables:

    __author__
    __version__
"""

from __future__ import annotations

__author__ = "Tobias Marczewski"
__version__ = '1.0.0'

import argparse
from collections.abc import Sequence


class PrintExampleFile(argparse.Action):
    """Argparse action to display an example orgmode file that can be parsed by deborg."""

    EXAMPLE_FILE: str = "# This is an example file for 'deborg'\n" +\
        "# deborg was written to parse orgmode list items of the general format:\n" +\
        "+ package-name {distro:release:tag1,tag2,tag3}\n" +\
        "\n" +\
        "# where each item can specify alternative packages separated by comma for the\n" +\
        "# same requirement\n" +\
        "+ package, package1 {distro1}, package2b {distro2:release_b}\n" +\
        "\n" +\
        "# for lines with several comma-separated packages deborg will at most return one\n" +\
        "# package, as they are seen as alternatives; if more than one package matches the\n" +\
        "# specifications an error will be reported.\n" +\
        "\n" +\
        "# As can be seen from the comment lines it is no problem to write more information into\n" +\
        "# the file, as long as the package lines (list items with either '-' or '+')\n" +\
        "# have the correct format. This additional information does not have to be in\n" +\
        "# comment format as we see below:\n" +\
        "\n" +\
        "* System relevant packages\n" +\
        "  The following packages I always want installed as they provide foo for baz...\n" +\
        "  + foo\n" +\
        "  + foo-two\n" +\
        "  # we can comment on the packages using description list items like that:\n" +\
        "  + baz, baz-alternative {distro1} :: for distro1 baz is missing in the repos\n" +\
        "                                      but baz-alternative provides the same\n" +\
        "				      functionality.\n" +\
        "  # and we can use tags like here\n" +\
        "  + apache {::server} :: this should be installed for any distro, but only on a\n" +\
        "                         server\n" +\
        "  + office-app-x {::desktop,laptop} :: this will be installed when 'desktop' or\n" +\
        "                                       'laptop' is passed as tag, hence not on a\n" +\
        "                                       server.\n" +\
        "\n" +\
        "* Email\n" +\
        "** general\n" +\
        "   # we can also use '-' as list bullet\n" +\
        "   - thunderbird\n" +\
        "\n" +\
        "** language support\n" +\
        "   - thunderbird-locale-xx {Ubuntu}, thunderbird-l10n-xx {Debian} :: language xx\n" +\
        "   - thunderbird-locale-zz {Ubuntu}, thunderbird-l10n-zz {Debian} :: language zz\n" +\
        ""

    def __init__(self, option_strings,
                 dest=argparse.SUPPRESS, default=argparse.SUPPRESS, **kwargs):
        super(PrintExampleFile, self).__init__(
            option_strings=option_strings, dest=dest, default=default, nargs=0, **kwargs)

    def __call__(self, parser, namespace, values, option_string=None):
        print(self.EXAMPLE_FILE)
        parser.exit()


class Examples(object):
    """
    Example usage for deborg using the file defined in PrintExampleFile.
    (deborg --example-file > examples.org)

    Running `deborg 'examples.org' Examples.input[i]` should produce `Examples.output[i]`
    """
    _input: list[str] = []
    _output: list[str] = []
    _n: int = 0

    def __init__(self):
        # 0
        self.add("'' ''",
                 "package foo foo-two baz thunderbird")
        # 1
        self.add("'distro1' ''",
                 "package1 foo foo-two baz-alternative thunderbird")
        # 2
        self.add("'distro2' 'release_b'",
                 "package2b foo foo-two baz thunderbird")
        # 3
        self.add("'' '' --tags=server",
                 "package foo foo-two baz apache thunderbird")
        # 4
        self.add("'distro1' '' --tags=desktop",
                 "package1 foo foo-two baz-alternative office-app-x thunderbird")
        # 5
        self.add("'distro1' '' --tags=desktop,server",
                 "package1 foo foo-two baz-alternative apache office-app-x thunderbird")
        # 6
        self.add("'Debian' ''",
                 "package foo foo-two baz thunderbird thunderbird-l10n-xx thunderbird-l10n-zz")
        self.add("'Ubuntu' ''",
                 "package foo foo-two baz thunderbird thunderbird-locale-xx thunderbird-locale-zz")
        # 8
        self.add("'distro' 'any' --sep='::'",
                 "package::foo::foo-two::baz::thunderbird")

    @property
    def input(self) -> Sequence[str]:
        return self._input

    @property
    def output(self) -> Sequence[str]:
        return self._output

    @property
    def n(self) -> int:
        return self._n

    def add(self, args: str, output: str):
        self._input.append(args)
        self._output.append(output)
        self._n += 1


def cli_parser() -> argparse.ArgumentParser:
    """Build a commandline argument parser for deborg"""
    indent: str = 2*" "
    disclaimer: str = "\ndisclaimer:\n\n" + \
        indent + f"Copyright (C) 2022 {__author__}\n" + \
        indent + "This program comes with ABSOLUTELY NO WARRANTY.\n" + \
        indent + "This is free software, and you are welcome to redistribute it under the terms\n" + \
        indent + "of the GNU General Public License version 3 or later.\n\n" + \
        indent + "For more information and bug reports please visit https://github.com/mtoboid/deborg\n"

    description: str = "\ndescription:\n\n" + \
        indent + "This program parses an orgfile with list entries (+ <item> or - <item>)\n" + \
        indent + "where each list item specifies one package or alternatives for the same package:\n" + \
        indent + " + package1, package1a {Ubuntu:18.04}, package1b {Debian:9}\n" + \
        indent + " + package2, package2a {Ubuntu}\n" + \
        indent + "...\n" + \
        indent + "where {<distro>:<release>} determine which package should be returned by deborg.\n" + \
        indent + "Also see '%(prog)s --example-file'.\n"

    ex: Examples = Examples()
    examples: str = "\nexamples:\n\n" + \
        indent + "For easier integration into scripts, it is advisable to use the exact strings returned\n" + \
        indent + "by a tool like 'lsb_release' for 'distro' and 'release'.\n" + \
        indent + "distro: lsb_release --short --id\n" + \
        indent + "release: lsb_release --short --release\n\n" + \
        indent + "$ deborg packages.org $(lsb_release --short --id) $(lsb_release --short --release)\n" + \
        indent + "$ package1 package2 package3\n\n" + \
        indent + "Using the example file 'deborg --example-file > examples.org' :\n\n"
    examples += ''.join(
        [f"{indent}$ deborg examples.org {ex.input[i]}\n{indent}$ {ex.output[i]}\n\n" for i in range(ex.n)])

    parser = argparse.ArgumentParser(
        prog="deborg",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        description="Extract Debian package information from an emacs .org file.",
        epilog="" + description + examples + disclaimer

    )
    parser.add_argument(
        "--example-file",
        help="Print an example orgmode file to stdout.",
        action=PrintExampleFile
    )
    parser.add_argument(
        "-v", "--version",
        help="Display the version of deborg.",
        action='version',
        version=f"%(prog)s {__version__}"
    )
    parser.add_argument(
        "orgfile",
        help="The .org file to parse.",
        type=str
    )
    parser.add_argument(
        "distro",
        help="Linux distribution for which to extract the packages, e.g. 'Debian', 'Ubuntu'...",
        type=str
    )
    parser.add_argument(
        "release",
        help="Release for which to extract the packages, e.g. '10', '11', '18.04'...",
        type=str
    )
    parser.add_argument(
        "-s", "--sep", default=" ",
        help="Separator used between package names in the returned array.",
        type=str
    )
    parser.add_argument(
        "-t", "--tags", default=None,
        dest="tags",
        help="Comma separated list of tags, no spaces. (tag1,tag2,tag3)",
        type=str
    )
    return parser
